R/derive_last_dose.R

```diff
@@ -83,143 +94,147 @@
 
   filter_ex <- assert_filter_cond(enquo(filter_ex), optional = TRUE)
   by_vars <- assert_vars(by_vars)
-  dose_start <- assert_symbol(enquo(dose_start))
-  dose_end <- assert_symbol(enquo(dose_end))
+  dose_id <- assert_vars(dose_id)
+  dose_date <- assert_symbol(enquo(dose_date))
   analysis_date <- assert_symbol(enquo(analysis_date))
   dataset_seq_var <- assert_symbol(enquo(dataset_seq_var))
-  new_var <- assert_symbol(enquo(new_var))
-  assert_logical_scalar(output_datetime)
-  assert_logical_scalar(check_dates_only)
-  stopifnot(is_quosures(traceability_vars) | is.null(traceability_vars))
+  single_dose_condition <- assert_filter_cond(enquo(single_dose_condition))
+  assert_varval_list(ex_keep_vars, optional = TRUE, accept_var = TRUE)
+  assert_varval_list(traceability_vars, optional = TRUE)
   assert_data_frame(dataset, quo_c(by_vars, analysis_date, dataset_seq_var))
-  assert_data_frame(dataset_ex, quo_c(by_vars, dose_start, dose_end))
+  assert_data_frame(dataset_ex, quo_c(by_vars, dose_date))
 
-  # by_vars converted to string
+  # vars converted to string
   by_vars_str <- vars2chr(by_vars)
+  dose_id_str <- vars2chr(dose_id)
 
-  # assumption for last dose derivation: start and end dates (datetimes) need to match
-  if (check_dates_only) {
-    check_cond <- summarise(dataset_ex,
-                            all_equal = all(as.Date(!!dose_start) == as.Date(!!dose_end)))
-  } else {
-    check_cond <- summarise(dataset_ex,
-                            all_equal = all(!!dose_start == !!dose_end))
+  # check if single_dose_condition is true for all EX records
+  single_dose_eval <- dataset_ex %>%
+    with(eval_tidy(quo_get_expr(single_dose_condition))) %>%
+    all()
+
+  if (!single_dose_eval){
+    stop("Specified single_dose_condition is not satisfied.")
   }
-  if (!check_cond$all_equal) {
-    stop(paste(
-      "Not all values of", as_string(quo_get_expr(dose_start)),
-      "are equal to", as_string(quo_get_expr(dose_end))
-    ))
+
+  # check if doses are unique based on dose_date and dose_id
+  signal_duplicate_records(dataset_ex, c(by_vars, dose_date, dose_id),
+  "Multiple doses exists for the same dose_date. Update dose_id to identify unique doses.")
+
+  # filter EX based on user-specified condition
+  if (!is.null(quo_get_expr(filter_ex))) {
+    dataset_ex <- dataset_ex %>%
+      filter_if(filter_ex)
   }
 
-  # run traceability if requested
+  # create traceability vars if requested
   if (!is.null(traceability_vars)) {
     trace_vars_str <- names(traceability_vars)
     dataset_ex <- mutate(dataset_ex, !!!traceability_vars)
   } else {
     trace_vars_str <- character(0)
   }
 
-  # select only a subset of columns
-  dataset_ex <- dataset_ex %>%
-    filter_if(filter_ex) %>%
-    select(!!!by_vars, !!dose_end, trace_vars_str)
+  # keep user-specified variables from EX, if no variables specified all EX variables are kept
+  if (is.null(ex_keep_vars)) {
+    ex_keep_vars_str <- colnames(dataset_ex)
```

Comment:
Hey @bundfussr I agree that this case could be removed. I wasn't around for the discussion but could we change the default for `ex_keep_vars` to `syms(colnames(dataset_ex)`? Then we could remove both the `if` and `else` conditions

Reply:
It would work but I would prefer
```
  if (!is.null(ex_keep_vars)) {
   dataset_ex <- dataset_ex %>% mutate(!!!ex_keep_vars) %>%
    select(!!!by_vars, !!!syms(dose_id_str), !!dose_date,
           !!!ex_keep_vars, !!!syms(trace_vars_str))
  }
```
because it is clearer and more efficient. Why should we call `mutate()` and `select()` if we do not want to change `dataset_ex` (in the case that `ex_keep_vars` is `NULL`)?

By the way, do we need the `mutate()` call?

Reply:
Of course, thanks that makes way more sense. I've changed it to that way around and removed the `mutate()`.